File: 02_Amazing/python/amazing.py
import random
import enum
from typing import List, Tuple
from dataclasses import dataclass

# Python translation by Frank Palazzolo - 2/2021


class Maze:
    def __init__(
        self,
        width: int,
        length: int,
    ):
        assert width >= 2 and length >= 2
        used: List[List[int]] = []
        walls: List[List[int]] = []
        for _ in range(length):
            used.append([0] * width)
            walls.append([0] * width)

        # Pick a random entrance, mark as used
        enter_col = random.randint(0, width - 1)
        used[0][enter_col] = 1

        self.used = used
        self.walls = walls
        self.enter_col = enter_col
        self.width = width
        self.length = length

    def add_exit(self) -> None:
        """Modifies 'walls' to add an exit to the maze."""
        col = random.randint(0, self.width - 1)
        row = self.length - 1
        self.walls[row][col] = self.walls[row][col] + 1

    def display(self) -> None:
        for col in range(self.width):
            if col == self.enter_col:
                print(".  ", end="")
            else:
                print(".--", end="")
        print(".")
        for row in range(self.length):
            print("I", end="")
            for col in range(self.width):
                if self.walls[row][col] < 2:
                    print("  I", end="")
                else:
                    print("   ", end="")
            print()
            for col in range(self.width):
                if self.walls[row][col] == 0 or self.walls[row][col] == 2:
                    print(":--", end="")
                else:
                    print(":  ", end="")
            print(".")


class Direction(enum.Enum):
    LEFT = 0
    UP = 1
    RIGHT = 2
    DOWN = 3


@dataclass
class Position:
    row: int
    col: int


# Give Exit directions nice names
EXIT_DOWN = 1
EXIT_RIGHT = 2


def main() -> None:
    welcome_header()
    width, length = get_maze_dimensions()
    maze = build_maze(width, length)
    maze.display()


def welcome_header() -> None:
    print(" " * 28 + "AMAZING PROGRAM")
    print(" " * 15 + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY")
    print()
    print()
    print()


def build_maze(width: int, length: int) -> Maze:
    """Build two 2D arrays."""
    #
    # used:
    #   Initially set to zero, unprocessed cells
    #   Filled in with consecutive non-zero numbers as cells are processed
    #
    # walls:
    #   Initially set to zero, (all paths blocked)
    #   Remains 0 if there is no exit down or right
    #   Set to 1 if there is an exit down
    #   Set to 2 if there is an exit right
    #   Set to 3 if there are exits down and right
    assert width >= 2 and length >= 2

    maze = Maze(width, length)
    position = Position(row=0, col=maze.enter_col)
    count = 2

    while count != width * length + 1:
        possible_dirs = get_possible_directions(maze, position)

        # If we can move in a direction, move and make opening
        if len(possible_dirs) != 0:
            position, count = make_opening(maze, possible_dirs, position, count)
        # otherwise, move to the next used cell, and try again
        else:
            while True:
                if position.col != width - 1:
                    position.col = position.col + 1
                elif position.row != length - 1:
                    position.row, position.col = position.row + 1, 0
                else:
                    position.row, position.col = 0, 0
                if maze.used[position.row][position.col] != 0:
                    break

    maze.add_exit()
    return maze


def make_opening(
    maze: Maze,
    possible_dirs: List[Direction],
    pos: Position,
    count: int,
) -> Tuple[Position, int]:
    """
    Attention! This modifies 'used' and 'walls'
    """
    direction = random.choice(possible_dirs)
    if direction == Direction.LEFT:
        pos.col = pos.col - 1
        maze.walls[pos.row][pos.col] = EXIT_RIGHT
    elif direction == Direction.UP:
        pos.row = pos.row - 1
        maze.walls[pos.row][pos.col] = EXIT_DOWN
    elif direction == Direction.RIGHT:
        maze.walls[pos.row][pos.col] = maze.walls[pos.row][pos.col] + EXIT_RIGHT
        pos.col = pos.col + 1
    elif direction == Direction.DOWN:
        maze.walls[pos.row][pos.col] = maze.walls[pos.row][pos.col] + EXIT_DOWN
        pos.row = pos.row + 1
    maze.used[pos.row][pos.col] = count
    count = count + 1
    return pos, count


def get_possible_directions(maze: Maze, pos: Position) -> List[Direction]:
    """
    Get a list of all directions that are not blocked.

    Also ignore hit cells that we have already processed
    """
    possible_dirs = list(Direction)
    if pos.col == 0 or maze.used[pos.row][pos.col - 1] != 0:
        possible_dirs.remove(Direction.LEFT)
    if pos.row == 0 or maze.used[pos.row - 1][pos.col] != 0:
        possible_dirs.remove(Direction.UP)
    if pos.col == maze.width - 1 or maze.used[pos.row][pos.col + 1] != 0:
        possible_dirs.remove(Direction.RIGHT)
    if pos.row == maze.length - 1 or maze.used[pos.row + 1][pos.col] != 0:
        possible_dirs.remove(Direction.DOWN)
    return possible_dirs


def get_maze_dimensions() -> Tuple[int, int]:
    while True:
        input_str = input("What are your width and length?")
        if input_str.count(",") == 1:
            width_str, length_str = input_str.split(",")
            width = int(width_str)
            length = int(length_str)
            if width > 1 and length > 1:
                break
        print("Meaningless dimensions. Try again.")
    return width, length


if __name__ == "__main__":
    main()

File: 02_Amazing/ruby/amazing.rb
# frozen_string_literal: true

DEBUG = !ENV['DEBUG'].nil?

require 'io/console' if DEBUG

# BASIC arrays are 1-based, unlike Ruby 0-based arrays,
# and this class simulates that. BASIC arrays are zero-filled,
# which is also done here. While we could easily update the
# algorithm to work with zero-based arrays, this class makes
# the problem easier to reason about, row or col 1 are the
# first row or column.
class BasicArrayTwoD
  def initialize(rows, cols)
    @val = Array.new(rows) { Array.new(cols, 0) }
  end

  def [](row, col = nil)
    if col
      @val[row - 1][col - 1]
    else
      @val[row - 1]
    end
  end

  def []=(row, col, n)
    @val[row - 1][col - 1] = n
  end

  def to_s(width: max_width, row_hilite: nil, col_hilite: nil)
    @val.map.with_index do |row, row_index|
      row.map.with_index do |val, col_index|
        if row_hilite == row_index + 1 && col_hilite == col_index + 1
          "[#{val.to_s.center(width)}]"
        else
          val.to_s.center(width + 2)
        end
      end.join
    end.join("\n")
  end

  def max_width
    @val.flat_map { |row| row.map { |val| val.to_s.length } }.sort.last
  end
end

class Maze
  EXIT_DOWN = 1
  EXIT_RIGHT = 2

  # Set up a constant hash for directions
  # The values represent the direction of the move as changes to row, col
  # and the type of exit when moving in that direction
  DIRECTIONS = {
    left: { row: 0, col: -1, exit: EXIT_RIGHT },
    up: { row: -1, col: 0, exit: EXIT_DOWN },
    right: { row: 0, col: 1, exit: EXIT_RIGHT },
    down: { row: 1, col: 0, exit: EXIT_DOWN }
  }.freeze

  attr_reader :width, :height, :used, :walls, :entry

  def initialize(width, height)
    @width = width
    @height = height

    @used = BasicArrayTwoD.new(height, width)
    @walls = BasicArrayTwoD.new(height, width)

    create
  end

  def draw
    # Print the maze
    draw_top(entry, width)
    (1..height - 1).each do |row|
      draw_row(walls[row])
    end
    draw_bottom(walls[height])
  end

  private

  def create
    # entry represents the location of the opening
    @entry = (rand * width).round + 1

    # Set up our current row and column, starting at the top and the locations of the opening
    row = 1
    col = entry
    c = 1
    used[row, col] = c # This marks the opening in the first row
    c += 1

    while c != width * height + 1 do
      debug walls, row, col
      # remove possible directions that are blocked or
      # hit cells that we have already processed
      possible_dirs = DIRECTIONS.reject do |dir, change|
        nrow = row + change[:row]
        ncol = col + change[:col]
        nrow < 1 || nrow > height || ncol < 1 || ncol > width || used[nrow, ncol] != 0
      end.keys

      # If we can move in a direction, move and make opening
      if possible_dirs.size != 0
        direction = possible_dirs.sample
        change = DIRECTIONS[direction] # pick a random direction
        if %i[left up].include?(direction)
          row += change[:row]
          col += change[:col]
          walls[row, col] = change[:exit]
        else
          walls[row, col] += change[:exit]
          row += change[:row]
          col += change[:col]
        end
        used[row, col] = c
        c = c + 1
      # otherwise, move to the next used cell, and try again
      else
        loop do
          if col != width
            col += 1
          elsif row != height
            row += 1
            col = 1
          else
            row = col = 1
          end
          break if used[row, col] != 0
          debug walls, row, col
        end
      end
    end

    # Add a random exit
    walls[height, (rand * width).round] += 1
  end

  def draw_top(entry, width)
    (1..width).each do |i|
      if i == entry
        print i == 1 ? '┏  ' : '┳  '
      else
        print i == 1 ? '┏━━' : '┳━━'
      end
    end

    puts '┓'
  end

  def draw_row(row)
    print '┃'
    row.each.with_index do |val, col|
      print val < 2 ? '  ┃' : '   '
    end
    puts
    row.each.with_index do |val, col|
      print val == 0 || val == 2 ? (col == 0 ? '┣━━' : '╋━━') : (col == 0 ? '┃  ' : '┫  ')
    end
    puts '┫'
  end

  def draw_bottom(row)
    print '┃'
    row.each.with_index do |val, col|
      print val < 2 ? '  ┃' : '   '
    end
    puts
    row.each.with_index do |val, col|
      print val == 0 || val == 2 ? (col == 0 ? '┗━━' : '┻━━') : (col == 0 ? '┗  ' : '┻  ')
    end
    puts '┛'
  end

  def debug(walls, row, col)
    return unless DEBUG

    STDOUT.clear_screen
    puts walls.to_s(row_hilite: row, col_hilite: col)
    sleep 0.1
  end
end

class Amazing
  def run
    draw_header

    width, height = ask_dimensions
    while width <= 1 || height <= 1
      puts "MEANINGLESS DIMENSIONS.  TRY AGAIN."
      width, height = ask_dimensions
    end

    maze = Maze.new(width, height)
    puts "\n" * 3
    maze.draw
  end

  def draw_header
    puts ' ' * 28 + 'AMAZING PROGRAM'
    puts ' ' * 15 + 'CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY'
    puts "\n" * 3
  end

  def ask_dimensions
    print 'WHAT ARE YOUR WIDTH AND HEIGHT? '
    width = gets.to_i
    print '?? '
    height = gets.to_i
    [width, height]
  end
end

Amazing.new.run